### PR TITLE
Convert RFF transform to estimators

### PR DESCRIPTION
**Pull request by Ivanidzo4ka**
This PR converts RFF transform to IEstimator and add PigstyExtension.

### CODE REVIEW

File: src/Microsoft.ML.Transforms/FourierDistributionSampler.cs [hunk 1]
```diff
@@ -9,6 +9,7 @@
 using Microsoft.ML.Runtime.Data;
 using Microsoft.ML.Runtime.Internal.Utilities;
 using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.EntryPoints;
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/FourierDistributionSampler.cs]: sort and get rid of Float

File: src/Microsoft.ML.Transforms/RffTransform.cs [hunk 2]
```diff
@@ -347,238 +297,442 @@ private static int RoundUp(int cflt, int cfltAlign)
...
+            int[] srcCols = new int[columns.Length];
+            for (int i = 0; i < columns.Length; i++)
+            {
+                if (!input.Schema.TryGetColumnIndex(ColumnPairs[i].input, out int srcCol))
+                    throw Host.ExceptSchemaMismatch(nameof(input), "input", ColumnPairs[i].input);
+                CheckInputColumn(input.Schema, i, srcCol);
+                srcCols[i] = srcCol;
+                activeColumns[srcCol] = true;
+            }
+            //IVAN: change namespace for ReservoirSamplerWithReplacement.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/RffTransform.cs]: >IVAN [](start = 14, length = 4)

indeed

File: src/Microsoft.ML.Transforms/RffTransform.cs [hunk 3]
```diff
@@ -224,115 +216,73 @@ private static VersionInfo GetVersionInfo()
                 loaderAssemblyName: typeof(RffTransform).Assembly.FullName);
         }
 
-        // These are parallel to Infos.
-        private readonly ColumnType[] _types;
         private readonly TransformInfo[] _transformInfos;
 
         private const string RegistrationName = "Rff";
```
Comment (artidoro) [inline src/Microsoft.ML.Transforms/RffTransform.cs]: Can you do nameof() instead?

File: test/Microsoft.ML.Tests/Transformers/RffTests.cs [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+            var est = data.MakeNewEstimator()
+                .Append(row => (
+                RffVectorFloat: row.VectorFloat.LowerVectorSizeWithRandomFourierTransformation(3, true),
+                Label: row.Label));
+
+            TestEstimatorCore(est.AsDynamic, data.AsDynamic);
+
+
+            var outputPath = GetOutputPath("Rff", "featurized.tsv");
+            using (var ch = Env.Start("save"))
```
Comment (artidoro) [inline test/Microsoft.ML.Tests/Transformers/RffTests.cs]: >ch [](start = 23, length = 2)

You mentioned needing ch.Done at the end of this?

File: src/Microsoft.ML.Transforms/RffTransform.cs [hunk 5]
```diff
@@ -347,238 +302,464 @@ private static int RoundUp(int cflt, int cfltAlign)
...
+        /// <param name="env">Host Environment.</param>
 
-            return
-                (ref VBuffer<Float> dst) =>
-                {
-                    getSrc(ref src);
-                    oneDimensionalVector.Values[0] = src;
-                    TransformFeatures(Host, ref oneDimensionalVector, ref dst, _transformInfos[iinfo], featuresAligned, productAligned);
-                };
+        /// <param name="inputColumn">Name of the output column.</param>
```
Comment (zeahmed) [inline src/Microsoft.ML.Transforms/RffTransform.cs]: >output [](start = 50, length = 6)

input.

File: src/Microsoft.ML.Transforms/RffTransform.cs [hunk 6]
```diff
@@ -347,238 +302,464 @@ private static int RoundUp(int cflt, int cfltAlign)
...
+        /// <param name="outputColumn">Name of the column to be transformed. If this is null '<paramref name="inputColumn"/>' will be used.</param>
```
Comment (zeahmed) [inline src/Microsoft.ML.Transforms/RffTransform.cs]: >Name of the column to be transformed. [](start = 39, length = 37)

This is the name of the output column not the column to be transformed, right?

File: src/Microsoft.ML.Transforms/RffTransform.cs [hunk 7]
```diff
@@ -347,238 +302,464 @@ private static int RoundUp(int cflt, int cfltAlign)
...
+        /// It maps input to a random low-dimensional feature space. It is useful when data has non-linear features, since the transform
+        /// is designed so that the inner products of the transformed data are approximately equal to those in the feature space of a user
+        /// speciﬁed shift-invariant kernel. With this transform, we are able to use linear methods (which are scalable) to approximate more complex kernel SVM models.
+        /// </summary>
+        /// <param name="input">The column to apply Random Fourier transfomration.</param>
+        /// <param name="newDim">Expected size of new vector.</param>
+        /// <param name="useSin">Create two features for every random Fourier frequency? (one for cos and one for sin) </param>
+        /// <param name="generator">Which kernel to use. (<see cref="GaussianFourierSampler"/> by default)</param>
+        /// <param name="seed">The seed of the random number generator for generating the new features. If not specified global random would be used.</param>
+        /// <returns></returns>
```
Comment (zeahmed) [inline src/Microsoft.ML.Transforms/RffTransform.cs]: >/// <returns></returns> [](start = 8, length = 23)

remove???

File: test/Microsoft.ML.Tests/Transformers/RffTests.cs [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+                new TestClass() { A = Enumerable.Range(0, 100).Select(x => (float)rand.NextDouble()).ToArray() }
+            };
+            var invalidData = ComponentCreation.CreateDataView(Env, new[] { new TestClassInvalidSchema { A = 1 }, new TestClassInvalidSchema { A = 1 } });
+            var validFitInvalidData = ComponentCreation.CreateDataView(Env, new[] { new TestClassBiggerSize { A = new float[200] }, new TestClassBiggerSize { A = new float[200] } });
+            var dataView = ComponentCreation.CreateDataView(Env, data);
+            var generator = new GaussianFourierSampler.Arguments();
+
+            var pipe = new RffEstimator(Env, new[]{
+                    new RffTransform.ColumnInfo("A", "RffA", 5, false),
+                    new RffTransform.ColumnInfo("A", "RffB", 10, true, new LaplacianFourierSampler.Arguments())
```
Comment (zeahmed) [inline test/Microsoft.ML.Tests/Transformers/RffTests.cs]: >"A" [](start = 48, length = 3)

"B"?

File: src/Microsoft.ML.Transforms/RffTransform.cs [hunk 9]
```diff
@@ -1,51 +1,47 @@
-// Licensed to the .NET Foundation under one or more agreements.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Transforms/RffTransform.cs]: >Licensed [](start = 3, length = 8)

Why are you doing this Ivan?

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Transforms/RffTransform.cs]: What can I say? I promise to be better, and not to do this again.

---
In reply to: [222843945](https://github.com/dotnet/machinelearning/pull/1122#discussion_r222843945) [](ancestors = 222843945)

Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Transforms/RffTransform.cs]: Ok, Fine, I forgive you this time, but better watch yourself!

---
In reply to: [222846814](https://github.com/dotnet/machinelearning/pull/1122#discussion_r222846814) [](ancestors = 222846814,222843945)

Comment (artidoro) [reply src/Microsoft.ML.Transforms/RffTransform.cs]: :joy: 

---
In reply to: [222846868](https://github.com/dotnet/machinelearning/pull/1122#discussion_r222846868) [](ancestors = 222846868,222846814,222843945)

File: src/Microsoft.ML.Transforms/FourierDistributionSampler.cs [hunk 10]
```diff
@@ -105,23 +111,25 @@ public void Save(ModelSaveContext ctx)
...
     {
-        public class Arguments
+        public class Arguments : IFourierDistributionSamplerFactory
         {
             [Argument(ArgumentType.AtMostOnce, HelpText = "a in the term exp(-a|x| / r). r is an estimate of the average intra-example L1 distance")]
-            public Float A = 1;
+            public float A = 1;
+
+            public IFourierDistributionSampler CreateComponent(IHostEnvironment env, float avgDist) => new LaplacianFourierSampler(env, this, avgDist);
         }
```
Comment (singlis) [inline src/Microsoft.ML.Transforms/FourierDistributionSampler.cs]: Is this a common pattern? Why not have Arguments not implement the IFourierDistributionSamplerFactory and instead have LaplacianFourierSampler implement this interface? Also there is no data from Arguments being used in the construction of the sampler, so its a bit odd to see this here.

File: src/Microsoft.ML.Transforms/RffTransform.cs [hunk 11]
```diff
@@ -213,126 +209,89 @@ private void InitializeFourierCoefficients(int rowSize, int colSize)
             + "shift-invariant kernel.";
 
         public const string LoaderSignature = "RffTransform";
+
         private static VersionInfo GetVersionInfo()
         {
             return new VersionInfo(
                 modelSignature: "RFF FUNC",
-                verWrittenCur: 0x00010001, // Initial
+                //verWrittenCur: 0x00010001, // Initial
```
Comment (singlis) [inline src/Microsoft.ML.Transforms/RffTransform.cs]: Is the commented out code left here intentionally to keep track of what changed? If not, I would delete.

File: src/Microsoft.ML.Transforms/FourierDistributionSampler.cs (hunk 10 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Transforms/FourierDistributionSampler.cs]: So you have two options (at least)
You can implement it in class.

```
IComponent component = new Class(new Class.Arguments()));
```
or you can do this:
```
IComponent component = new Class.Arguments();
```
Which one would you prefer?

If you worried about properties of Argument class not been used, I actually pass into Class constructor "this" object, which is argument class, and it got used in constructor

---
In reply to: [222859347](https://github.com/dotnet/machinelearning/pull/1122#discussion_r222859347) [](ancestors = 222859347)

File: src/Microsoft.ML.Transforms/RffTransform.cs (hunk 11 above)
Comment (Ivanidzo4ka) [reply src/Microsoft.ML.Transforms/RffTransform.cs]: This is kinda our pattern how we handle situations with bumping version.
looks weird, but you have some hopes what previous version got started from 10001 and not something else.

---
In reply to: [222860021](https://github.com/dotnet/machinelearning/pull/1122#discussion_r222860021) [](ancestors = 222860021)

**Pull request merged**